### PR TITLE
Allow named color palettes to be used by all sources.

### LINKED ISSUE #716: Generalize named colors and palettes
Currently, the GDAL and mapnik tilesources allow using named palettes from the `palettable` library instead of specifying an array of colors.  This should be generalized to the base tile source.  Additionally, it would be handle if a named color could be used as a short-hand to specify a range of black to the named color.

Additionally, it would be nice to use palette names from matplotlib that AREN'T in palettable (palettable has old values).  Additionally, matplotlib has named colors that aren't in `PIL.ImageColor.colormap` (PIL just has the css colors).  It the ideal, we'd support these colors, too.

We don't want to add matplotlib as a dependency, as it is heavy.

### PR DESCRIPTION
**Pull request by manthey**
This also increase the ability to handle css and named colors.

This generalized and supercedes #714.  This closes #716.

### CODE REVIEW

**Review by banesullivan — COMMENTED** [review]
This is excellent! However, I'm having some trouble getting this to work:

No problem:
```py
>>> large_image.open('/Users/bane/Desktop/TC_NG_SFBay_US_Geo.tif', style=json.dumps({"band": 1, "palette": ["#000", "#f00"]}))
<large_image_source_gdal.GDALFileTileSource object at 0x7fa138fe3430>
>>>
```

Problem:

```py
>>> large_image.open('/Users/bane/Desktop/TC_NG_SFBay_US_Geo.tif', style=json.dumps({"band": 1, "palette": "viridis"}))
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/Users/bane/Software/kw/large_image/large_image/tilesource/__init__.py", line 140, in open
    return getTileSource(*args, **kwargs)
  File "/Users/bane/Software/kw/large_image/large_image/tilesource/__init__.py", line 128, in getTileSource
    return getTileSourceFromDict(AvailableTileSources, *args, **kwargs)
  File "/Users/bane/Software/kw/large_image/large_image/tilesource/__init__.py", line 113, in getTileSourceFromDict
    sourceName = getSourceNameFromDict(availableSources, pathOrUri, *args, **kwargs)
  File "/Users/bane/Software/kw/large_image/large_image/tilesource/__init__.py", line 98, in getSourceNameFromDict
    if availableSources[sourceName].canRead(pathOrUri, *args, **kwargs):
  File "/Users/bane/Software/kw/large_image/large_image/tilesource/base.py", line 2360, in canRead
    cls(path, *args, **kwargs)
  File "/Users/bane/Software/kw/large_image/large_image/cache_util/cache.py", line 193, in __call__
    instance = super().__call__(*args, **kwargs)
  File "/Users/bane/Software/kw/large_image/sources/gdal/large_image_source_gdal/__init__.py", line 156, in __init__
    self._setDefaultStyle()
  File "/Users/bane/Software/kw/large_image/sources/gdal/large_image_source_gdal/__init__.py", line 227, in _setDefaultStyle
    bstyle['palette'] = self.getHexColors(bstyle['palette'])
  File "/Users/bane/Software/kw/large_image/sources/gdal/large_image_source_gdal/__init__.py", line 358, in getHexColors
    return ['#%02X%02X%02X%02X' % tuple(clr) for clr in palette]
  File "/Users/bane/Software/kw/large_image/sources/gdal/large_image_source_gdal/__init__.py", line 358, in <listcomp>
    return ['#%02X%02X%02X%02X' % tuple(clr) for clr in palette]
TypeError: %X format: an integer is required, not numpy.float64
>>>
```

**Comment by manthey** [comment]
Should be fixed and tested now.

File: large_image/tilesource/utilities.py [hunk 1]
```diff
@@ -398,3 +401,71 @@ def nearPowerOfTwo(val1, val2, tolerance=0.02):
...
+            import matplotlib
+
+            palette = (
+                ['#0000', matplotlib.colors.to_hex(value)]
+                if value in matplotlib.colors.get_named_colors_mapping()
+                else matplotlib.cm.get_cmap(value).colors)
+        except (ImportError, ValueError):
+            pass
+    if palette is None:
+        raise TileSourceError('Value cannot be used as a color palette.')
```
[the remark this answers is not in the document]
Comment (manthey) [reply large_image/tilesource/utilities.py:470]: My thought was that it prevents the tile source from working, but I'll change it to a `ValueError`.

**Comment by banesullivan** [comment]
Would it make sense to fail back to a known value? For example, `jet` is a valid colormap if `colorcet` is installed. I'd prefer for this `getPaletteColors` to raise a `ValueError` if the named colormap does not exist and for `large_image.open` to safely fail back to a known palette issuing a warning if a bad palette name is passed

This might not be desired by everyone though...

**Comment by banesullivan** [comment]
Also could we implement an `isValidPalette()` function to let downstream packages quickly check a palette?

**Comment by manthey** [comment]
> Also could we implement an `isValidPalette()` function to let downstream packages quickly check a palette?

Sure -- this would just be 
```
try:
    large_image.tilesource.utilities.getPaletteColors(<value>)
    return True
except ValueError:
    return False
```

**Comment by banesullivan** [comment]
Could we also have a `getAvailablePalettes()` method to list all available palettes - this would be particularly useful when making front-end that let you select from a list of colormaps. For example what I have here: https://tileserver.banesullivan.com/?band=2&min=&max=&palette=magma

**Comment by banesullivan** [comment]
> I'd rather not. It seems like passing something like #fffff should throw, rather than produce something unexpected.

Fair enough - leaving it as is works for me

**Comment by manthey** [comment]
> Could we also have a `getAvailablePalettes()` method to list all available palettes - this would be particularly useful when making front-end that let you select from a list of colormaps. For example what I have here: https://tileserver.banesullivan.com/?band=2&min=&max=&palette=magma

Should such a function actually be called `getAvailableNamedPalettes()`, since it is basically a list of all of the palettable palettes and matplotlib palettes.  Or, since you can pass a single color to get a palette, would such a function return named colors, too?

**Comment by banesullivan** [comment]
> getAvailableNamedPalettes

I was *just* thinking that -- yes

> would such a function return named colors, too?

I think so, I think anything named should be returned (e.g., "red", etc.)

**Comment by manthey** [comment]
> > getAvailableNamedPalettes

Let's push off adding `getAvailableNamedPalettes` into a separate PR.

**Comment by manthey** [comment]
@banesullivan There is a now a `isValidPalette` method.

**Review by banesullivan — APPROVED** [review]

**Pull request merged**